File: Examples/SimpleAppObjectiveC/SimpleAppObjectiveC/ViewController.m
//
//  ViewController.m
//  SimpleAppObjectiveC
//
//  Created by Daehoon Kim on 7/24/25.
//

#import "ViewController.h"
#import "Constants.h"

@interface ViewController ()

@property (nonatomic, strong) UILabel *titleLabel;
@property (nonatomic, strong) UILabel *statusLabel;
@property (nonatomic, strong) UIButton *predictButton;
@property (nonatomic, strong) UIButton *logEventButton;
@property (nonatomic, strong) UILabel *discountStatusLabel;
@property (nonatomic, strong) UILabel *resultLabel;
@property (nonatomic, strong) NSTimer *statusCheckTimer;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    [self setupUI];
    [self setupMonetaiSDK];
    [self setupNotifications];
    
    // Log app launch event
    [[MonetaiSDK shared] logEventWithEventName:@"app_in" params:nil];
}

- (void)dealloc {
    [[NSNotificationCenter defaultCenter] removeObserver:self];
    [self.statusCheckTimer invalidate];
}

#pragma mark - UI Setup

- (void)setupUI {
    self.view.backgroundColor = [UIColor systemBackgroundColor];
    
    // Title Label
    self.titleLabel = [[UILabel alloc] init];
    self.titleLabel.text = @"MonetaiSDK Demo";
    self.titleLabel.textColor = [UIColor labelColor];
    self.titleLabel.font = [UIFont boldSystemFontOfSize:28];
    self.titleLabel.textAlignment = NSTextAlignmentCenter;
    self.titleLabel.translatesAutoresizingMaskIntoConstraints = NO;
    [self.view addSubview:self.titleLabel];
    
    // Status Label
    self.statusLabel = [[UILabel alloc] init];
    self.statusLabel.text = @"SDK Status: Initializing...";
    self.statusLabel.textColor = [UIColor secondaryLabelColor];
    self.statusLabel.font = [UIFont systemFontOfSize:16];
    self.statusLabel.textAlignment = NSTextAlignmentCenter;
    self.statusLabel.translatesAutoresizingMaskIntoConstraints = NO;
    [self.view addSubview:self.statusLabel];
    
    // Predict Button
    self.predictButton = [UIButton buttonWithType:UIButtonTypeSystem];
    [self.predictButton setTitle:@"Predict Purchase" forState:UIControlStateNormal];
    self.predictButton.titleLabel.font = [UIFont systemFontOfSize:18 weight:UIFontWeightMedium];
    self.predictButton.backgroundColor = [UIColor systemBlueColor];
    [self.predictButton setTitleColor:[UIColor whiteColor] forState:UIControlStateNormal];
    self.predictButton.layer.cornerRadius = 10;
    self.predictButton.translatesAutoresizingMaskIntoConstraints = NO;
    [self.predictButton addTarget:self action:@selector(predictButtonTapped) forControlEvents:UIControlEventTouchUpInside];
    [self.view addSubview:self.predictButton];
    
    // Log Event Button
    self.logEventButton = [UIButton buttonWithType:UIButtonTypeSystem];
    [self.logEventButton setTitle:@"Log Event" forState:UIControlStateNormal];
    self.logEventButton.titleLabel.font = [UIFont systemFontOfSize:16 weight:UIFontWeightMedium];
    self.logEventButton.backgroundColor = [UIColor systemGreenColor];
    [self.logEventButton setTitleColor:[UIColor whiteColor] forState:UIControlStateNormal];
    self.logEventButton.layer.cornerRadius = 8;
    self.logEventButton.translatesAutoresizingMaskIntoConstraints = NO;
    [self.logEventButton addTarget:self action:@selector(logEventButtonTapped) forControlEvents:UIControlEventTouchUpInside];
    [self.view addSubview:self.logEventButton];
    
    // Discount Status Label
    self.discountStatusLabel = [[UILabel alloc] init];
    self.discountStatusLabel.text = @"Discount: None";
    self.discountStatusLabel.textColor = [UIColor secondaryLabelColor];
    self.discountStatusLabel.font = [UIFont systemFontOfSize:14];
    self.discountStatusLabel.textAlignment = NSTextAlignmentCenter;
    self.discountStatusLabel.translatesAutoresizingMaskIntoConstraints = NO;
    [self.view addSubview:self.discountStatusLabel];
    
    // Result Label
    self.resultLabel = [[UILabel alloc] init];
    self.resultLabel.text = @"Tap buttons to test SDK functionality";
    self.resultLabel.textColor = [UIColor secondaryLabelColor];
    self.resultLabel.font = [UIFont systemFontOfSize:14];
    self.resultLabel.textAlignment = NSTextAlignmentCenter;
    self.resultLabel.numberOfLines = 0;
    self.resultLabel.translatesAutoresizingMaskIntoConstraints = NO;
    [self.view addSubview:self.resultLabel];
    
    [self setupConstraints];
}

- (void)setupConstraints {
    [NSLayoutConstraint activateConstraints:@[
        // Title Label
        [self.titleLabel.centerXAnchor constraintEqualToAnchor:self.view.centerXAnchor],
        [self.titleLabel.topAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.topAnchor constant:50],
        
        // Status Label
        [self.statusLabel.centerXAnchor constraintEqualToAnchor:self.view.centerXAnchor],
        [self.statusLabel.topAnchor constraintEqualToAnchor:self.titleLabel.bottomAnchor constant:20],
        
        // Predict Button
        [self.predictButton.centerXAnchor constraintEqualToAnchor:self.view.centerXAnchor],
        [self.predictButton.centerYAnchor constraintEqualToAnchor:self.view.centerYAnchor constant:-30],
        [self.predictButton.widthAnchor constraintEqualToConstant:200],
        [self.predictButton.heightAnchor constraintEqualToConstant:50],
        
        // Log Event Button
        [self.logEventButton.centerXAnchor constraintEqualToAnchor:self.view.centerXAnchor],
        [self.logEventButton.topAnchor constraintEqualToAnchor:self.predictButton.bottomAnchor constant:20],
        [self.logEventButton.widthAnchor constraintEqualToConstant:150],
        [self.logEventButton.heightAnchor constraintEqualToConstant:40],
        
        // Discount Status Label
        [self.discountStatusLabel.centerXAnchor constraintEqualToAnchor:self.view.centerXAnchor],
        [self.discountStatusLabel.topAnchor constraintEqualToAnchor:self.logEventButton.bottomAnchor constant:30],
        
        // Result Label
        [self.resultLabel.centerXAnchor constraintEqualToAnchor:self.view.centerXAnchor],
        [self.resultLabel.topAnchor constraintEqualToAnchor:self.discountStatusLabel.bottomAnchor constant:20],
        [self.resultLabel.leadingAnchor constraintEqualToAnchor:self.view.leadingAnchor constant:20],
        [self.resultLabel.trailingAnchor constraintEqualToAnchor:self.view.trailingAnchor constant:-20]
    ]];
}

#pragma mark - MonetaiSDK Setup

- (void)setupMonetaiSDK {
    // Set up discount info change callback
    __weak typeof(self) weakSelf = self;
    [MonetaiSDK shared].onDiscountInfoChangeCallback = ^(id _Nullable discountInfo) {
        dispatch_async(dispatch_get_main_queue(), ^{
            [weakSelf handleDiscountInfoChange:discountInfo];
        });
    };
    
    // Start checking SDK status periodically
    [self startSDKStatusCheck];
    
    // Configure paywall
    [self configurePaywall];
}

- (void)configurePaywall {
    // Create features array using the correct initializer
    Feature *feature1 = [[Feature alloc] initWithTitle:@"Unlimited Access" 
                                          description:@"Use all premium features without limits" 
                                         isPremiumOnly:NO];
    
    Feature *feature2 = [[Feature alloc] initWithTitle:@"Advanced Analytics" 
                                          description:@"AI-powered insights" 
                                         isPremiumOnly:YES];
    
    NSArray<Feature *> *features = @[feature1, feature2];
    
    // Create paywall configuration using the new Options-based approach
    PaywallConfigOptions *options = [[PaywallConfigOptions alloc] init];
    options.features = features;
    
    PaywallConfig *config = [[PaywallConfig alloc] initWithDiscountPercent:30
                                                              regularPrice:@"$99.99"
                                                           discountedPrice:@"$69.99"
                                                                     locale:@"en"
                                                                      style:PaywallStyleHighlightBenefits
                                                                   options:options];
    
    // Set up callbacks using the correct property names
    __weak typeof(self) weakSelf = self;
    config.onPurchase = ^(void (^close)(void)) {
        NSLog(@"[SimpleAppObjectiveC] onPurchase called");
        dispatch_async(dispatch_get_main_queue(), ^{
            weakSelf.resultLabel.text = @"🛒 Purchase flow started";
            weakSelf.resultLabel.textColor = [UIColor systemBlueColor];
        });
        
        dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.1 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
            NSLog(@"[SimpleAppObjectiveC] calling close()");
            close();
            [[MonetaiSDK shared] setSubscriptionStatus:YES];
        });
    };
    
    config.onTermsOfService = ^{
        NSLog(@"[SimpleAppObjectiveC] onTermsOfService called");
        dispatch_async(dispatch_get_main_queue(), ^{
            [weakSelf presentAlertWithTitle:@"Terms of Service" message:@"Open Terms of Service URL"];
        });
    };
    
    config.onPrivacyPolicy = ^{
        NSLog(@"[SimpleAppObjectiveC] onPrivacyPolicy called");
        dispatch_async(dispatch_get_main_queue(), ^{
            [weakSelf presentAlertWithTitle:@"Privacy Policy" message:@"Open Privacy Policy URL"];
        });
    };
    
    // Configure the paywall
    [[MonetaiSDK shared] configurePaywallWithConfig:config];
    [[MonetaiSDK shared] setSubscriptionStatus:NO];
}

- (void)setupNotifications {
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(sdkInitialized)
                                                 name:kMonetaiSDKInitializedNotification
                                               object:nil];
    
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(sdkInitializationFailed:)
                                                 name:kMonetaiSDKInitializationFailedNotification
                                               object:nil];
}

- (void)sdkInitialized {
    dispatch_async(dispatch_get_main_queue(), ^{
        [self updateSDKStatus];
        self.resultLabel.text = @"✅ SDK initialized successfully!";
        self.resultLabel.textColor = [UIColor systemGreenColor];
    });
}

- (void)sdkInitializationFailed:(NSNotification *)notification {
    dispatch_async(dispatch_get_main_queue(), ^{
        [self updateSDKStatus];
        NSError *error = notification.object;
        if (error) {
            self.resultLabel.text = [NSString stringWithFormat:@"❌ SDK initialization failed: %@", error.localizedDescription];
        } else {
            self.resultLabel.text = @"❌ SDK initialization failed";
        }
        self.resultLabel.textColor = [UIColor systemRedColor];
    });
}

- (void)startSDKStatusCheck {
    // Check status immediately
    [self updateSDKStatus];
    
    // Check status every 1 second until initialized
    self.statusCheckTimer = [NSTimer scheduledTimerWithTimeInterval:1.0 target:self selector:@selector(checkSDKStatus) userInfo:nil repeats:YES];
}

- (void)checkSDKStatus {
    BOOL isInitialized = [[MonetaiSDK shared] getInitialized];
    [self updateSDKStatus];
    
    // Stop timer when SDK is initialized
    if (isInitialized) {
        [self.statusCheckTimer invalidate];
        self.statusCheckTimer = nil;
    }
}

- (void)updateSDKStatus {
    BOOL isInitialized = [[MonetaiSDK shared] getInitialized];
    self.statusLabel.text = isInitialized ? @"SDK Status: ✅ Ready" : @"SDK Status: ⏳ Initializing...";
    self.statusLabel.textColor = isInitialized ? [UIColor systemGreenColor] : [UIColor systemOrangeColor];
    
    // Enable/disable buttons based on initialization status
    self.predictButton.enabled = isInitialized;
    self.logEventButton.enabled = isInitialized;
    
    if (isInitialized) {
        self.predictButton.alpha = 1.0;
        self.logEventButton.alpha = 1.0;
    } else {
        self.predictButton.alpha = 0.5;
        self.logEventButton.alpha = 0.5;
    }
}

- (void)handleDiscountInfoChange:(AppUserDiscount *)discountInfo {
    if (discountInfo) {
        NSDate *now = [NSDate date];
        NSDate *endTime = discountInfo.endedAt;
        
        if ([now compare:endTime] == NSOrderedAscending) {
            // Discount is valid
            NSDateFormatter *formatter = [[NSDateFormatter alloc] init];
            formatter.dateStyle = NSDateFormatterShortStyle;
            formatter.timeStyle = NSDateFormatterShortStyle;
            NSString *formattedDate = [formatter stringFromDate:endTime];
            
            self.discountStatusLabel.text = [NSString stringWithFormat:@"Discount: ✅ Active (Expires: %@)", formattedDate];
            self.discountStatusLabel.textColor = [UIColor systemGreenColor];
            
            // Update result label
            self.resultLabel.text = @"🎉 Discount is now active!\nSpecial offer available.";
            self.resultLabel.textColor = [UIColor systemGreenColor];
        } else {
            // Discount expired
            self.discountStatusLabel.text = @"Discount: ❌ Expired";
            self.discountStatusLabel.textColor = [UIColor systemRedColor];
            
            // Update result label
            self.resultLabel.text = @"Discount offer has expired";
            self.resultLabel.textColor = [UIColor systemRedColor];
        }
    } else {
        // No discount
        self.discountStatusLabel.text = @"Discount: None";
        self.discountStatusLabel.textColor = [UIColor secondaryLabelColor];
        
        // Update result label
        self.resultLabel.text = @"No active discount offers";
        self.resultLabel.textColor = [UIColor secondaryLabelColor];
    }
}

#pragma mark - Button Actions

- (void)predictButtonTapped {
    [[MonetaiSDK shared] predictWithCompletion:^(PredictResponse * _Nullable result, NSError * _Nullable error) {
        if (error) {
            dispatch_async(dispatch_get_main_queue(), ^{
                self.resultLabel.text = [NSString stringWithFormat:@"❌ Prediction failed: %@", error.localizedDescription];
                self.resultLabel.textColor = [UIColor systemRedColor];
            });
            NSLog(@"Prediction failed: %@", error.localizedDescription);
        } else {
            NSLog(@"Prediction result: %@", result.predictionString ?: @"None");
            NSLog(@"Test group: %@", result.testGroupString ?: @"None");

            if ([result.predictionString isEqualToString:@"non-purchaser"]) {
                // When predicted as non-purchaser, offer discount
                NSLog(@"Predicted as non-purchaser - discount can be applied");
            } else if ([result.predictionString isEqualToString:@"purchaser"]) {
                // When predicted as purchaser
                NSLog(@"Predicted as purchaser - discount not needed");
            }
            
            dispatch_async(dispatch_get_main_queue(), ^{
                self.resultLabel.text = @"✅ Prediction completed - check console for details";
                self.resultLabel.textColor = [UIColor systemGreenColor];
                
                // Show alert with prediction result
                UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"Purchase Prediction"
                                                                               message:[NSString stringWithFormat:@"Prediction: %@\nTest Group: %@",
                                                                                        result.predictionString ?: @"None",
                                                                                        result.testGroupString ?: @"None"]
                                                                        preferredStyle:UIAlertControllerStyleAlert];
                
                UIAlertAction *okAction = [UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault handler:nil];
                [alert addAction:okAction];
                [self presentViewController:alert animated:YES completion:nil];
            });
        }
    }];
}

- (void)logEventButtonTapped {
    // Log a sample event with parameters
    NSDictionary *params = @{
        @"button": @"test_button",
        @"screen": @"main"
    };
    
    [[MonetaiSDK shared] logEventWithEventName:@"button_click" params:params];
    
    dispatch_async(dispatch_get_main_queue(), ^{
        self.resultLabel.text = @"✅ Event logged: button_click\nParameters: button=test_button, screen=main";
        self.resultLabel.textColor = [UIColor systemGreenColor];
    });
    
    NSLog(@"Event logged: button_click");
}

#pragma mark - Helper Methods

- (void)presentAlertWithTitle:(NSString *)title message:(NSString *)message {
    UIAlertController *alert = [UIAlertController alertControllerWithTitle:title
                                                                   message:message
                                                            preferredStyle:UIAlertControllerStyleAlert];
    
    UIAlertAction *okAction = [UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault handler:nil];
    [alert addAction:okAction];
    
    dispatch_async(dispatch_get_main_queue(), ^{
        UIViewController *presenter = self.presentedViewController ?: [self topMostViewController] ?: self;
        [presenter presentViewController:alert animated:YES completion:nil];
    });
}

- (UIViewController *)topMostViewController {
    UIWindow *keyWindow = nil;
    
    if (@available(iOS 13.0, *)) {
        for (UIWindowScene *windowScene in [UIApplication sharedApplication].connectedScenes) {
            if (windowScene.activationState == UISceneActivationStateForegroundActive) {
                for (UIWindow *window in windowScene.windows) {
                    if (window.isKeyWindow) {
                        keyWindow = window;
                        break;
                    }
                }
                if (keyWindow) break;
            }
        }
    }
    
    if (!keyWindow) {
        keyWindow = [UIApplication sharedApplication].windows.firstObject;
    }
    
    UIViewController *topVC = keyWindow.rootViewController;
    while (topVC.presentedViewController) {
        topVC = topVC.presentedViewController;
    }
    
    return topVC;
}

@end

File: Examples/SimpleApp/SimpleApp/ViewController.swift
//
//  ViewController.swift
//  SimpleApp
//
//  Created by Daehoon Kim on 7/23/25.
//

import UIKit
import MonetaiSDK

class ViewController: UIViewController {

    // MARK: - UI Elements
    private let titleLabel = UILabel()
    private let statusLabel = UILabel()
    private let predictButton = UIButton(type: .system)
    private let logEventButton = UIButton(type: .system)
    private let discountStatusLabel = UILabel()
    private let resultLabel = UILabel()
    private var discountBannerView: DiscountBannerView?

    override func viewDidLoad() {
        super.viewDidLoad()
        
        setupUI()
        setupMonetaiSDK()
        setupNotifications()
        
        // Log app launch event
        Task {
            await MonetaiSDK.shared.logEvent(eventName: "app_in")
        }
    }
    
    deinit {
        NotificationCenter.default.removeObserver(self)
    }
    
    // MARK: - UI Setup
    private func setupUI() {
        view.backgroundColor = UIColor.systemBackground
        
        // Title Label
        titleLabel.text = "MonetaiSDK Demo"
        titleLabel.textColor = UIColor.label
        titleLabel.font = UIFont.boldSystemFont(ofSize: 28)
        titleLabel.textAlignment = .center
        titleLabel.translatesAutoresizingMaskIntoConstraints = false
        view.addSubview(titleLabel)
        
        // Status Label
        statusLabel.text = "SDK Status: Initializing..."
        statusLabel.textColor = UIColor.secondaryLabel
        statusLabel.font = UIFont.systemFont(ofSize: 16)
        statusLabel.textAlignment = .center
        statusLabel.translatesAutoresizingMaskIntoConstraints = false
        view.addSubview(statusLabel)
        
        // Predict Button
        predictButton.setTitle("Predict Purchase", for: .normal)
        predictButton.titleLabel?.font = UIFont.systemFont(ofSize: 18, weight: .medium)
        predictButton.backgroundColor = UIColor.systemBlue
        predictButton.setTitleColor(UIColor.white, for: .normal)
        predictButton.layer.cornerRadius = 10
        predictButton.translatesAutoresizingMaskIntoConstraints = false
        predictButton.addTarget(self, action: #selector(predictButtonTapped), for: .touchUpInside)
        view.addSubview(predictButton)
        
        // Log Event Button
        logEventButton.setTitle("Log Event", for: .normal)
        logEventButton.titleLabel?.font = UIFont.systemFont(ofSize: 16, weight: .medium)
        logEventButton.backgroundColor = UIColor.systemGreen
        logEventButton.setTitleColor(UIColor.white, for: .normal)
        logEventButton.layer.cornerRadius = 8
        logEventButton.translatesAutoresizingMaskIntoConstraints = false
        logEventButton.addTarget(self, action: #selector(logEventButtonTapped), for: .touchUpInside)
        view.addSubview(logEventButton)

		// Removed manual Paywall controls per requirement
        
        // Discount Status Label
        discountStatusLabel.text = "Discount: None"
        discountStatusLabel.textColor = UIColor.secondaryLabel
        discountStatusLabel.font = UIFont.systemFont(ofSize: 14)
        discountStatusLabel.textAlignment = .center
        discountStatusLabel.translatesAutoresizingMaskIntoConstraints = false
        view.addSubview(discountStatusLabel)
        
        // Result Label
        resultLabel.text = "Tap buttons to test SDK functionality"
        resultLabel.textColor = UIColor.secondaryLabel
        resultLabel.font = UIFont.systemFont(ofSize: 14)
        resultLabel.textAlignment = .center
        resultLabel.numberOfLines = 0
        resultLabel.translatesAutoresizingMaskIntoConstraints = false
        view.addSubview(resultLabel)
        
        setupConstraints()
    }
    
    private func setupConstraints() {
        NSLayoutConstraint.activate([
            // Title Label
            titleLabel.centerXAnchor.constraint(equalTo: view.centerXAnchor),
            titleLabel.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor, constant: 50),
            
            // Status Label
            statusLabel.centerXAnchor.constraint(equalTo: view.centerXAnchor),
            statusLabel.topAnchor.constraint(equalTo: titleLabel.bottomAnchor, constant: 20),
            
            // Predict Button
            predictButton.centerXAnchor.constraint(equalTo: view.centerXAnchor),
            predictButton.centerYAnchor.constraint(equalTo: view.centerYAnchor, constant: -30),
            predictButton.widthAnchor.constraint(equalToConstant: 200),
            predictButton.heightAnchor.constraint(equalToConstant: 50),
            
            // Log Event Button
            logEventButton.centerXAnchor.constraint(equalTo: view.centerXAnchor),
            logEventButton.topAnchor.constraint(equalTo: predictButton.bottomAnchor, constant: 20),
            logEventButton.widthAnchor.constraint(equalToConstant: 150),
            logEventButton.heightAnchor.constraint(equalToConstant: 40),

			// Manual Paywall buttons removed
            
			// Discount Status Label
			discountStatusLabel.centerXAnchor.constraint(equalTo: view.centerXAnchor),
			discountStatusLabel.topAnchor.constraint(equalTo: logEventButton.bottomAnchor, constant: 30),
            
            // Result Label
            resultLabel.centerXAnchor.constraint(equalTo: view.centerXAnchor),
            resultLabel.topAnchor.constraint(equalTo: discountStatusLabel.bottomAnchor, constant: 20),
            resultLabel.leadingAnchor.constraint(equalTo: view.leadingAnchor, constant: 20),
            resultLabel.trailingAnchor.constraint(equalTo: view.trailingAnchor, constant: -20)
        ])
    }
    
    // MARK: - MonetaiSDK Setup
    private func setupMonetaiSDK() {
        // Set up discount info change callback
        MonetaiSDK.shared.onDiscountInfoChange = { [weak self] discountInfo in
            DispatchQueue.main.async {
                self?.handleDiscountInfoChange(discountInfo)
            }
        }
        
        // Start checking SDK status periodically
        startSDKStatusCheck()

		// Configure paywall
		configurePaywall()
    }

	private func configurePaywall() {
		let features = [
			Feature(title: "Unlimited Access", description: "Use all premium features without limits"),
			Feature(title: "Advanced Analytics", description: "AI-powered insights", isPremiumOnly: true)
		]
		let config = PaywallConfig(
			discountPercent: 30,
			regularPrice: "$99.99",
			discountedPrice: "$69.99",
			locale: "en",
			features: features,
			style: .highlightBenefits
		)

		config.onPurchase = { [weak self] close in
			print("[SimpleApp] onPurchase called")
			self?.resultLabel.text = "🛒 Purchase flow started"
			self?.resultLabel.textColor = UIColor.systemBlue
			DispatchQueue.main.asyncAfter(deadline: .now() + 0.1) {
				print("[SimpleApp] calling close()")
				close()
			}
		}
		config.onTermsOfService = { [weak self] in
			print("[SimpleApp] onTermsOfService called")
			self?.presentAlert(title: "Terms of Service", message: "Open Terms of Service URL")
		}
		config.onPrivacyPolicy = { [weak self] in
			print("[SimpleApp] onPrivacyPolicy called")
			self?.presentAlert(title: "Privacy Policy", message: "Open Privacy Policy URL")
		}

		MonetaiSDK.shared.configurePaywall(config)
	}
    
    private func setupNotifications() {
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(sdkInitialized),
            name: .monetaiSDKInitialized,
            object: nil
        )
        
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(sdkInitializationFailed),
            name: .monetaiSDKInitializationFailed,
            object: nil
        )
    }
    
    @objc private func sdkInitialized() {
        DispatchQueue.main.async {
            self.updateSDKStatus()
            self.resultLabel.text = "✅ SDK initialized successfully!"
            self.resultLabel.textColor = UIColor.systemGreen
        }
    }
    
    @objc private func sdkInitializationFailed(_ notification: Notification) {
        DispatchQueue.main.async {
            self.updateSDKStatus()
            if let error = notification.object as? Error {
                self.resultLabel.text = "❌ SDK initialization failed: \(error.localizedDescription)"
                self.resultLabel.textColor = UIColor.systemRed
            }
        }
    }
    
    private func startSDKStatusCheck() {
        // Check status immediately
        updateSDKStatus()
        
        // Check status every 1 second until initialized
        Timer.scheduledTimer(withTimeInterval: 1.0, repeats: true) { [weak self] timer in
            guard let self = self else {
                timer.invalidate()
                return
            }
            
            let isInitialized = MonetaiSDK.shared.getInitialized()
            self.updateSDKStatus()
            
            // Stop timer when SDK is initialized
            if isInitialized {
                timer.invalidate()
            }
        }
    }
    
    private func updateSDKStatus() {
        let isInitialized = MonetaiSDK.shared.getInitialized()
        statusLabel.text = isInitialized ? "SDK Status: ✅ Ready" : "SDK Status: ⏳ Initializing..."
        statusLabel.textColor = isInitialized ? UIColor.systemGreen : UIColor.systemOrange
        
        // Enable/disable buttons based on initialization status
        predictButton.isEnabled = isInitialized
        logEventButton.isEnabled = isInitialized
        
        if isInitialized {
            predictButton.alpha = 1.0
            logEventButton.alpha = 1.0
        } else {
            predictButton.alpha = 0.5
            logEventButton.alpha = 0.5
        }
    }
    
	private func handleDiscountInfoChange(_ discountInfo: AppUserDiscount?) {
		if let discount = discountInfo {
			let now = Date()
			let endTime = discount.endedAt
			if now < endTime {
				discountStatusLabel.text = "Discount: ✅ Active (Expires: \(endTime.formatted(date: .abbreviated, time: .shortened)))"
				discountStatusLabel.textColor = UIColor.systemGreen
			} else {
				discountStatusLabel.text = "Discount: ❌ Expired"
				discountStatusLabel.textColor = UIColor.systemRed
			}
		} else {
			discountStatusLabel.text = "Discount: None"
			discountStatusLabel.textColor = UIColor.secondaryLabel
		}
	}
    
    // MARK: - Button Actions
    @objc private func predictButtonTapped() {
        Task {
            do {
                let result = try await MonetaiSDK.shared.predict()

                print("Prediction result:", result.prediction?.stringValue ?? "None")
                print("Test group:", result.testGroup?.stringValue ?? "None")

                if result.prediction == .nonPurchaser {
                    // When predicted as non-purchaser, offer discount
                    print("Predicted as non-purchaser - discount can be applied")
                } else if result.prediction == .purchaser {
                    // When predicted as purchaser
                    print("Predicted as purchaser - discount not needed")
                }
                
                await MainActor.run {
                    resultLabel.text = "✅ Prediction completed - check console for details"
                    resultLabel.textColor = UIColor.systemGreen
						
						// Show alert with prediction result (present from top-most VC safely)
						self.presentAlert(
							title: "Purchase Prediction",
							message: "Prediction: \(result.prediction?.stringValue ?? "None")\nTest Group: \(result.testGroup?.stringValue ?? "None")"
						)
                }
                
            } catch {
                await MainActor.run {
                    resultLabel.text = "❌ Prediction failed: \(error.localizedDescription)"
                    resultLabel.textColor = UIColor.systemRed
                }
                print("Prediction failed:", error)
            }
        }
    }
    
    @objc private func logEventButtonTapped() {
        Task {
            // Log a sample event with parameters
            await MonetaiSDK.shared.logEvent(
                eventName: "button_click",
                params: ["button": "test_button", "screen": "main"]
            )
            
            await MainActor.run {
                resultLabel.text = "✅ Event logged: button_click\nParameters: button=test_button, screen=main"
                resultLabel.textColor = UIColor.systemGreen
            }
            
            print("Event logged: button_click")
        }
    }

	private func presentAlert(title: String, message: String) {
		let alert = UIAlertController(title: title, message: message, preferredStyle: .alert)
		alert.addAction(UIAlertAction(title: "OK", style: .default))
		DispatchQueue.main.async {
			let presenter = self.presentedViewController ?? self.topMostViewController() ?? self
			presenter.present(alert, animated: true)
		}
	}

	private func topMostViewController() -> UIViewController? {
		if let windowScene = UIApplication.shared.connectedScenes
			.compactMap({ $0 as? UIWindowScene })
			.first(where: { $0.activationState == .foregroundActive }) {
			if let keyWindow = windowScene.windows.first(where: { $0.isKeyWindow }) ?? windowScene.windows.first {
				var topVC = keyWindow.rootViewController
				while let presented = topVC?.presentedViewController { topVC = presented }
				return topVC
			}
		}
		if let window = UIApplication.shared.windows.first(where: { $0.isKeyWindow }) ?? UIApplication.shared.windows.first {
			var topVC = window.rootViewController
			while let presented = topVC?.presentedViewController { topVC = presented }
			return topVC
		}
		return nil
	}
}

